Packet++/src/TcpSorter.cpp

```diff
@@ -0,0 +1,585 @@
+#define LOG_MODULE PacketLogModuleTcpSorter
+
+#include "TcpSorter.h"
+#include "TcpLayer.h"
+#include "IPv4Layer.h"
+#include "IPv6Layer.h"
+#include "PacketUtils.h"
+#include "IpAddress.h"
+#include "Logger.h"
+#include <sstream>
+#include <random>
+#include <vector>
+#include <algorithm>
+#if defined(WIN32) || defined(PCAPPP_MINGW_ENV) //for using ntohl, ntohs, etc.
+#include <winsock2.h>
+#elif LINUX
+#include <in.h> //for using ntohl, ntohs, etc.
+#elif MAC_OS_X || FREEBSD
+#include <arpa/inet.h> //for using ntohl, ntohs, etc.
+#endif
+
+#define PURGE_FREQ_SECS 1
+
+namespace pcpp
+{
+
+TcpSorter::TcpSorter(
+							OnTcpPacketReady onPacketReadyCallback,
+							OnTcpPacketMissing onPacketMissingCallback,
+							void* userCookie,
+							const TcpSorterConfiguration &config)
+{
+	m_isClosed = false;
+	m_OnPacketReadyCallback = onPacketReadyCallback;
+	m_OnPacketMissingCallback = onPacketMissingCallback;
+	m_UserCookie = userCookie;
+	// configuration value
+	m_MaxNumCapturedPacket = config.maxNumCapturedPacket;
+	m_MaxIdleTimeout = config.maxIdleTimeout;
+	m_MaxNumInactiveConnScan = config.maxNumInactiveConnScan;
+	m_MaxSegmentLifeTime = config.maxSegmentLifeTime;
+	m_CleanUpInactiveConnPeriod = config.cleanUpInactiveConnPeriod;
+	m_ShouldIncludeEmptySegments = config.shouldIncludeEmptySegments;
+	// reset clean up timer
+	m_LastCleanupTime = time(NULL);
+}
+
+TcpSorter::~TcpSorter()
+{
+	closeAllConnections();
+}
+
+void TcpSorter::sortPacket(SPRawPacket spRawPacket)
+{
+	// stop sorting if it is closed
+	if(m_isClosed)
+	{
+		return;
+	}
+	// parse the packet
+	Packet packet(spRawPacket.get(), false);
```

Comment:
You can make a small optimization if it is applicable: `Packet packet(spRawPacket.get(), false, TCP)`